Fix: Import never finishes

largestgithubuseronearth opened the pull request:
Fixes #270

This is caused by a race condition in the importer class. `GLib.idle_add` calls the `pipeline_advanced_callback` too often, so the `n_pipelines_done` counter becomes greater than the total game pipelines. This usually isn't a factor because the source tasks usually finish before the pipelines, but when they don't, the importer can never be "finished" importing. So importing is permanently stalled even though all the work was successful.

From my testing, this fixes the issue and doesn't seem to break anything else. But I don't really understand the comment which says it updates a widget. What widget? It doesn't seem to have a clear effect on the progress bar.

largestgithubuseronearth commented:
I added a couple logging statements to the counter increments to hopefully catch this mistake in future.

Let me know of any changes you'd like to see. Thanks.

kra-mo commented:
Thank you ^^

The issue is that `pipeline_advanced_callback` calls `progress_changed_callback`, which calls `update_progressbar`, which then calls `GtkProgressBar.set_fraction`. But the `idle_add` could be added later in the chain.

kra-mo commented:
For the record, the dialog not closing is partially a libadw thing, some of it got resolved in #302

largestgithubuseronearth commented:
I tested the progressbar fraction with and without `idle_add`. In practice, the `idle_add` is just redundantly calling `update_progressbar()` when no new work has been done.

IMO `idle_add` should be removed altogether as it is redundant and prone to causing errors. Connecting the pipeline "advanced" signal to the `pipeline_advanced_callback` is entirely sufficient to update the progressbar. And logically, it makes sense that it should only update when progress has been made and there is something to report.

Let me know what you think.

With `idle_add`:
![Screenshot From 2024-10-31 13-25-38](https://github.com/user-attachments/assets/9613c2ff-d2bc-406a-9a9c-0ded6070e2d8)

Without: 
![Screenshot From 2024-10-31 13-29-07](https://github.com/user-attachments/assets/9a7e9e21-cae5-4a4c-b026-434d7d28e3ed)

kra-mo commented:
`idle_add` is there so that the app doesn't crash. GTK widgets are not thread-safe, so they have to be updated in the main thread.

largestgithubuseronearth commented:
It seems the simplest solution seems to be to have two separate callbacks for the pipeline "advanced" signal, one for the counter and one for the progressbar.

On another note, it seems there's something else also causing unpredictable crashes during import. I've been trying to figure out the cause because it makes it hard to tell if any change I try is causing it more or less often. Importing as soon as app launches and spamming import seem to cause it more often. But I've had it happen a few times now without trying to cause it.

largestgithubuseronearth commented:
I think this is a pretty okay solution. I took your watchdog and gave it full responsibility over updating the dialog and the import clean-up which completely the GUI actions from the threaded code. No need to tie the threads back to the main path with `idle_add`.

I never found the cause of the crashing but it seems to be a separate issue (it also occurs in `main` right now). It seems to be somewhere in the middle of the importing because it prevents importing from finishing. You can cause it by holding down the import action shortcut.

Let me know what you think or if changes need to be made. Thanks.

kra-mo commented:
I tested the branch, but it does not seem to work with auto-import on. The import gets stuck. The watchdog is needed.

kra-mo commented:
Can you also please provide a summary of all of your changes? It gets hard to follow over the comments and commits.

largestgithubuseronearth commented:
> I tested the branch, but it does not seem to work with auto-import on. The dialog never closes. The watchdog is needed. For the record, this is because of a libadwaita bug. Not a lot you can do about it.

I don't know what bug you're referring to. Can you post a link? I haven't been able to replicate that but I'll look into it.

> Can you also please provide a summary of all of your changes? It gets hard to follow over the comments and commits.

Sure, my bad. 

Summary:
 - Instead of the pipelines calling `progress_changed_callback()` when they've finished/progressed, the watchdog is responsible for updating the progressbar and it calls the `import_callback()` when everything is finished.
 - Removed `progress_changed_callback()` since watchdog now bears this responsibility.
 - Renamed `import_callback()` -> `finish_import()` for clarity of purpose.
 - Renamed `__watchdog()` -> `monitor_import()` for clarity of purpose.
 - Reorganized methods into groups for "GUI" and "Threaded" so it's a bit easier to tell if a method is in a thread.

largestgithubuseronearth commented:
> It's not in a release, so it's not reported by anyone else.

What do you mean it's not in a release? Adw or Cartridges?

> I can reproduce it myself just by turning auto-import on.

I can't reproduce it with that setting on. It works as intended for me so I'm not sure yet what the issue is. I need more details.

Can you outline what the bug is exactly? Under what circumstances does it happen? What is your workaround and why does it work?

> I don't know if it is reported in libadwaita or not, but it is the reason there was a timeout for closing the dialog in the first place.

What's confusing me is that the timeout still exists. I didn't remove the watchdog or the timeout, I just have it call the cleanup method when the import is `finished` and that method force closes the dialog. I only removed the `progress_changed_callback()`. 

What is the important part of the puzzle here? Is it the timeout or retrying the `force_close()` method multiple times or something else?

kra-mo commented:
> What do you mean it's not in a release? Adw or Cartridges?

Cartridges

> Can you outline what the bug is exactly? Under what circumstances does it happen? What is your workaround and why does it work?

My guess is if you force close a dialog too fast after opening it, it stays open.

> What is the important part of the puzzle for the timeout? Is it retrying the force_close() method multiple times or running the code in the main thread or something else?

GTK code should always be run in the main thread anyway.

The fix is to wait before closing the dialog. But the time you have to wait is seemingly inconsistent, hence why I tried multiple times until it succeeded.

kra-mo commented:
But regardless, we just can’t rely on force close actually working, so we need to retry it until it does 🤷

largestgithubuseronearth commented:
Okay. That's not a lot to go on but I'll see what I can do.

In the meantime, consider reporting it to libadw.

largestgithubuseronearth commented:
I added a while loop to the `finish_import()` method that will retry it until it closes properly.

Give it a try when you have time and let me know if it fixes it.

kra-mo commented:
> I added a while loop to the finish_import() method that will keep force closing the dialog until it closes properly.

This does not work. It simply makes the app freeze. Just do it periodically, there is a reason I did it that way originally.

kra-mo commented:
Also, please remove the comments I mentioned in the review.

kra-mo commented:
> In the meantime, consider reporting it to libadw.

I barely have the motivation and energy to maintain my own projects nowadays. Reporting a bug upstream is a lot of effort, if you want to make it actually valuable, sorry.

largestgithubuseronearth commented:
> This does not work. It simply makes the app freeze. Just do it periodically, there is a reason I did it that way originally.

Sure.

> Also, please remove the comments I mentioned in the review.

What review?

largestgithubuseronearth commented on cartridges/importer/importer.py:
```diff
@@ -69,6 +70,8 @@ def __init__(self) -> None:
         self.game_pipelines = set()
         self.sources = set()
 
+    """PUBLIC"""
```
How would you do it then? Or you want me to remove them?

I added them because it was hard for me to track what code could be touched in a thread. Otherwise you need to rely on name or following the code again and again.

kra-mo replied:
You can add comments to those (though I wouldn't do all uppercase), but private/public is indicated by the function name itself so I'd remove those.

largestgithubuseronearth replied:
Sure

largestgithubuseronearth commented:
Okay I've kept watchdog as it was. `monitor_import()` is the same but I re-added the watchdog as it was as.

I also removed the public and private comments.

Thanks for your feedback. Let me know if that fixes it.

kra-mo commented:
Seems to work. So can you confirm that this resolves #270?

largestgithubuseronearth commented:
That's good. Yes, I'm pretty confident.

largestgithubuseronearth commented:
Great. Thanks for your help.

kra-mo commented:
Thanks for the PR.

kra-mo commented:
Ah, that’s good to know.

largestgithubuseronearth commented:
I look back at this pr and I cringe. I'm sorry for having been so annoying to work with. I'm often a rube
